### PR TITLE
FIxes #126   Self-signed certificate issue

### LINKED ISSUE #126: Add validation for custom/self-signed certificates
Zulip 1.5.1 running on Ubuntu 16.04.2 in hyper-V 
the ubuntu machine is called zulip with a self-signed certificate
the settings.py has as external_host = 'zulip'
we only use it internally without access from the outside of the LAN
in the browser https://zulip works or https://localipaddress
in the electron app however this doesn't work (nor with the IP)
image

### PR DESCRIPTION
**Pull request by Lplenka**
Currently our app doesn't support servers with self-signed certificates.

This PR fixes that.
When a user tries to connect a server with a self-signed certificate, a dialog box appears asking for permission. If permission is granted the app loads the server. Also, that server is added to the list of `CertifiedURL`  in app's backend, so the user doesn't face the dialog box in future if he tries to connect the same server but only initially once.

dialogbox

Any suggestions on improving this PR are welcome :smile:

### CODE REVIEW

**Comment by akashnimare** [comment]
This looks good. @timabbott is it possible for you to cross check it? As of now, I don't have access to a Linux machine so I'm not sure whether this fix will work or not :thinking: .

**Comment by akashnimare** [comment]
@Lplenka there is lot of code shared between `pref.js` and `domain.js`. We could reuse it.

**Comment by Lplenka** [comment]
> @Lplenka there is lot of code shared between pref.js and domain.js. We could reuse it.

Yes, they share a lot in common, only the HTML elements they use with are different. You can suggest me the changes, I will add them. But most probably, in the long run we won't be having a `pref.js` since the app will support multipe servers so server switching related files will be deleted eventually.

File: app/renderer/index.html [hunk 1]
```diff
@@ -20,8 +20,7 @@
                             Zulip Server URL
                         </label>
                         <input type="text" id="url"  autofocus="autofocus" spellcheck="false" placeholder="Server URL">
-                        <button type="submit" id="main" class="btn-primary btn-large" value="Submit"
-                                onclick="addDomain();">Connect</button>
```
Comment (Lplenka) [inline app/renderer/index.html]: Removed reductant call to addDomain()

**Comment by timabbott** [comment]
To test this, I'd recommend getting a cheap VM and installing the Zulip production environment following  http://zulip.readthedocs.io/en/latest/prod-install.html, using the SSL certificate documentation.  

I think @feorlen would probably be able to cross-test this pretty quickly, since she's pretty practiced with the installer (though I think she's on a trip this week, so may be slow to reply) :)

**Comment by feorlen** [comment]
Sure, I can have a look when I'm home early next week. I'm away without a computer at the moment.

**Comment by akashnimare** [comment]
This partially fixes the self-signed cert problem. If a user enters a server which is self-signed, app will ask for permission to load that URL -



@Lplenka merging this. Next step would be select the appropriate certificate from keychain using [select-client-certificate](https://github.com/electron/electron/blob/master/docs/api/app.md#event-select-client-certificate).

**Comment by Lplenka** [comment]
> @Lplenka merging this. Next step would be select the appropriate certificate from keychain using select-client-certificate.

Yes, surely will work on that.

**Comment by feorlen** [comment]
I tried to test this, following the directions in the zulip-electron README.md (since now it's been merged to master) but install failed. I've never installed the desktop app before, so I'm not sure what should happen. This is on a ubuntu VM that previously was running the webapp (dev), so I would think it would have all the expected stuff. 

 Do you still need someone to verify this against a running server? 

Here's what happened:

first attempt at `npm install` fails here:
```
npm ERR! Linux 3.13.0-24-generic
npm ERR! argv "/usr/local/nvm/versions/node/v6.6.0/bin/node" "/usr/local/nvm/versions/node/v6.6.0/bin/npm" "install"
npm ERR! node v6.6.0
npm ERR! npm  v3.10.3
npm ERR! code ELIFECYCLE

npm ERR! electron@1.4.15 postinstall: `node install.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the electron@1.4.15 postinstall script 'node install.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the electron package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node install.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs electron
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls electron
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/feorlen/zulip-electron/npm-debug.log
```

Here's what's in the debug log:
```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/nvm/versions/node/v6.6.0/bin/node',
1 verbose cli   '/usr/local/nvm/versions/node/v6.6.0/bin/npm',
1 verbose cli   'start' ]
2 info using npm@3.10.3
3 info using node@v6.6.0
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle zulip@0.5.10~prestart: zulip@0.5.10
6 silly lifecycle zulip@0.5.10~prestart: no script for prestart, continuing
7 info lifecycle zulip@0.5.10~start: zulip@0.5.10
8 verbose lifecycle zulip@0.5.10~start: unsafe-perm in lifecycle true
9 verbose lifecycle zulip@0.5.10~start: PATH: /usr/local/nvm/versions/node/v6.6.
0/lib/node_modules/npm/bin/node-gyp-bin:/home/feorlen/zulip-electron/node_module
s/.bin:/usr/local/nvm/versions/node/v6.6.0/bin:/usr/local/nvm/versions/node/v6.6
.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/
usr/local/games
10 verbose lifecycle zulip@0.5.10~start: CWD: /home/feorlen/zulip-electron
11 silly lifecycle zulip@0.5.10~start: Args: [ '-c', 'electron ./app/main' ]
12 info lifecycle zulip@0.5.10~start: Failed to exec start script
13 verbose stack Error: zulip@0.5.10 start: `electron ./app/main`
13 verbose stack spawn ENOENT
13 verbose stack     at ChildProcess.<anonymous> (/usr/local/nvm/versions/node/v
6.6.0/lib/node_modules/npm/lib/utils/spawn.js:33:16)
13 verbose stack     at emitTwo (events.js:106:13)
13 verbose stack     at ChildProcess.emit (events.js:191:7)
13 verbose stack     at maybeClose (internal/child_process.js:877:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_proc
ess.js:226:5)
14 verbose pkgid zulip@0.5.10
15 verbose cwd /home/feorlen/zulip-electron
16 error Linux 3.13.0-24-generic
17 error argv "/usr/local/nvm/versions/node/v6.6.0/bin/node" "/usr/local/nvm/ver
sions/node/v6.6.0/bin/npm" "start"
18 error node v6.6.0
19 error npm  v3.10.3
20 error file sh
21 error code ELIFECYCLE
22 error errno ENOENT
23 error syscall spawn
24 error zulip@0.5.10 start: `electron ./app/main`
24 error spawn ENOENT
25 error Failed at the zulip@0.5.10 start script 'electron ./app/main'.
25 error Make sure you have the latest version of node.js and npm installed.
25 error If you do, this is most likely a problem with the zulip package,
25 error not with npm itself.
25 error Tell the author that this fails on your system:
25 error     electron ./app/main
25 error You can get information on how to open an issue for this project with:
25 error     npm bugs zulip
25 error Or if that isn't available, you can get their info via:
25 error     npm owner ls zulip
25 error There is likely additional logging output above.
26 verbose exit [ 1, true ]
```

After running provision on the Zulip install already there, repeating npm install also shows this error:

```
feorlen-VirtualBox[feorlen]:~/zulip-electron$ npm install

> electron@1.4.15 postinstall /home/feorlen/zulip-electron/node_modules/electron
> node install.js

/home/feorlen/zulip-electron/node_modules/electron/install.js:46
  throw err
  ^

Error: ENOENT: no such file or directory, lstat '/home/feorlen/zulip-electron/node_modules/electron/dist/resources'
    at Error (native)
```

**Comment by Lplenka** [comment]
@feorlen Thanks for your help on this. 

> Do you still need someone to verify this against a running server?

It would be better if we get positive reviews from all those you are using the self-certified zulip server. So please test it.

Please upgrade  nodejs  to `6.9.4` or more recent.
You should remove node_modules using command `rm -rf node_modules && rm -rf app/node_modules` and try a fresh install after upgrading nodejs.
Here is a [discussion](https://chat.zulip.org/#narrow/stream/electron/topic/server.20url) on a similar problem.

**Comment by feorlen** [comment]
Hi @Lplenka I'm trying to follow the directions in the conversation you linked, but I'm confused. Is there another way to install nodejs 6.9.4 without first building nvm from source? apt-get on ubuntu only installs 6.6.0.
```
feorlen-VirtualBox[feorlen]:~/zulip-electron$ sudo apt-get install npm
Reading package lists... Done
Building dependency tree       
Reading state information... Done
npm is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
feorlen-VirtualBox[feorlen]:~/zulip-electron$ which npm
/usr/local/nvm/versions/node/v6.6.0/bin/npm
```

When I try to build/install nvm using the install script from its repo, I get a message that nvm is already installed and I should update it directly from git into /usr/local/nvm.
```
=> nvm is already installed in /usr/local/nvm, trying to update using git
=> Failed to update nvm, run 'git fetch' in /usr/local/nvm yourself.
```

Why would I need a git repository for anything in /usr/local? Is this normal?

**Comment by feorlen** [comment]
@Lplenka I'll do that, as updating v6.10.2 is giving me the same error as when I tried v6.9.4. See you in chat!

**Comment by feorlen** [comment]
I don't understand why node gave me so much trouble, but @Lplenka and I got it working. I'm prompted to confirm the self-signed cert. Yay!

**Comment by Lplenka** [comment]
@feorlen Thanks! :smile:

**Pull request merged**